components/timezone/src/ids.rs

```diff
@@ -0,0 +1,670 @@
+                    let removed_byte = unsafe { string.as_mut_vec().pop() };
+                    if let Some(removed_byte) = removed_byte {
+                        if !removed_byte.is_ascii() {
+                            debug_assert!(false, "non-ASCII removed byte: {removed_byte}");
+                            // If we get here for some reason, `string` is not in a valid state,
+                            // so to be extra safe, we can clear it.
+                            string.clear();
+                            return None;
+                        }
+                    } else {
+                        debug_assert!(false, "could not remove another byte");
+                        return None;
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// A mapper that supplements [`TimeZoneIdMapper`] with additional data to improve the performance
+/// of canonical IANA ID lookup.
+#[derive(Debug, Clone)]
+pub struct TimeZoneIdMapperWithFastCanonicalization<I> {
+    inner: I,
+    data: DataPayload<Bcp47ToIanaMapV1Marker>,
+}
+
+impl TimeZoneIdMapperWithFastCanonicalization<TimeZoneIdMapper> {
+    /// Creates a new [`TimeZoneIdMapperWithFastCanonicalization`] using compiled data.
+    ///
+    /// See [`TimeZoneIdMapperWithFastCanonicalization`] for an example.
+    ///
+    /// ✨ *Enabled with the `compiled_data` Cargo feature.*
+    ///
+    /// [📚 Help choosing a constructor](icu_provider::constructors)
+    #[cfg(feature = "compiled_data")]
+    pub fn new() -> Self {
+        const _: () = assert!(
+            crate::provider::Baked::SINGLETON_TIME_ZONE_IANA_TO_BCP47_V2.bcp47_ids_checksum
+                == crate::provider::Baked::SINGLETON_TIME_ZONE_BCP47_TO_IANA_V1.bcp47_ids_checksum,
+        );
+        Self {
+            inner: TimeZoneIdMapper {
+                data: DataPayload::from_static_ref(
+                    crate::provider::Baked::SINGLETON_TIME_ZONE_IANA_TO_BCP47_V2,
+                ),
+            },
+            data: DataPayload::from_static_ref(
+                crate::provider::Baked::SINGLETON_TIME_ZONE_BCP47_TO_IANA_V1,
+            ),
+        }
+    }
+
+    icu_provider::gen_any_buffer_data_constructors!(locale: skip, options: skip, error: TimeZoneError,
+        #[cfg(skip)]
+        functions: [
+            new,
+            try_new_with_any_provider,
+            try_new_with_buffer_provider,
+            try_new_unstable,
+            Self,
+        ]
+    );
+
+    #[doc = icu_provider::gen_any_buffer_unstable_docs!(UNSTABLE, Self::new)]
+    pub fn try_new_unstable<P>(provider: &P) -> Result<Self, TimeZoneError>
+    where
+        P: DataProvider<IanaToBcp47MapV2Marker> + DataProvider<Bcp47ToIanaMapV1Marker> + ?Sized,
+    {
+        let mapper = TimeZoneIdMapper::try_new_unstable(provider)?;
+        Self::try_new_with_mapper_unstable(provider, mapper)
+    }
+}
+
+impl<I> TimeZoneIdMapperWithFastCanonicalization<I>
```

Comment:
Happy for feedback on the name of this think. `TimeZoneIdMapperWithFastCanonicalization`